WIP RFC Replace usage of USEPKG by USEMODULE.

cladmi opened the pull request:
### Contribution description

This addresses the inconsistent use of USEPKG mentioned in #6668.
Currently USEPKG is required to tell the build system to take `pkg/PKG_NAME` into account.

It was argued that it makes the user aware he is using a package as he is required to put USEPKG, but packages are being pulled as dependencies for some cpu. And also, some packages, like `lwip` are implemented by several modules and doing `USEMODULE = lwip_ipv6` triggers the inclusion of the package transparently.

This PR tries to implement the proposed solution

> I propose making USEMODULE += implicitly add to USEPKG, iff pkg/ exists.

The build system modification is done around these changes:

``` diff
diff --git a/Makefile.dep b/Makefile.dep
-# include package dependencies
--include $(USEPKG:%=$(RIOTPKG)/%/Makefile.dep)
+# process package dependencies
+-include $(ALL_PKGS:%=$(RIOTPKG)/%/Makefile.dep)

diff --git a/Makefile.include b/Makefile.include
 # process dependencies
 include $(RIOTBASE)/Makefile.dep
+# deduce packages from USEMODULE
+USEPKG := $(filter $(ALL_PKGS), $(USEMODULE))
```

and fixing packages so `pkg/*/Makefile.dep` can be included even if the module is not required. Like done in the normal Makefile.dep anyway.

This goes in the direction of making packages self contained like #9003 

## API CHANGE

It requires all packages `Makefile.dep` to be protected by the `ifneq` pattern.
It also require to remove all `USEPKG` in applications (which should be asserted right now).


## Implementation

The commits have been designed to be incremental and always be in a working state. But this was not tested.
I count on murdock to help finding if there are issues and fixing them in the right place.

## TODO

* [ ] Get maintainers feedback on the solution
* [ ] Check if the updated documentation is enough ! ! !
* [ ] Move all easy packages dependencies to `pkg/*/Makefile.dep` (this could be done in subsequent PRs also)
* [ ] check if if really all packages dependencies can be safely moved or document them


### Issues/PRs references

pkg: USEPKG vs USEMODULE #6668

cladmi committed:
Makefile.include: list all packages directories

It is assumed they are in 'RIOTPKG/pkgname' and that there is a module
named 'pkgname'.

cladmi committed:
Makefile.dep: deduce USEPKG from modules in USEMODULE

Prepare for specifying dependencies to packages only by putting them in
`USEMODULE`.

USEPKG would be an internal only variable similar to DIRS.

cladmi committed:
Makefile.dep: replace dependencies set as USEPKG to USEMODULE

Prepare for specifying dependencies to packages only by putting them in
`USEMODULE`.

cladmi committed:
boards/cpu: replace dependencies set as USEPKG to USEMODULE

Prepare for specifying dependencies to packages only by putting them in
`USEMODULE`.

cladmi committed:
pkg/doc.txt: remove outdated information on INCLUDE

INCLUDE is now set by the packages pkg/<pkg_name>/Makefile.include file.

cladmi committed:
pkg/relic: remove wrong information about variable definition

There is no such thing as a required order to define this.

cladmi committed:
TODO: MAYBE UPDATE THE DOC.TXT SENTENCE pkg: replace dependencies set as USEPKG to USEMODULE

Prepare for secifying dependencies to packages only by putting them in
`USEMODULE`.

cladmi committed:
sys: replace dependencies checking as USEPKG to USEMODULE

Prepare for secifying dependencies to packages only by putting them in
`USEMODULE`.

cladmi committed:
tests: replace dependencies set as USEPKG to USEMODULE

Prepare for secifying dependencies to packages only by putting them in
`USEMODULE`.

cladmi committed:
examples: replace dependencies set as USEPKG to USEMODULE

Prepare for secifying dependencies to packages only by putting them in
`USEMODULE`.

cladmi committed:
make: deprecate dependency handling from USEPKG to USEMODULE

Replace declaring dependency of packages using USEPKG to using USEMODULE.
Setting USEPKG to the list of used packages will be done by the build system.

https://github.com/RIOT-OS/RIOT/issues/6668

Currently, if a package has sub modules, like pkg_name_sub_module it would still
require to have in the main Makefile.dep a dependency from pkg_name_sub_module
to pkg_name module.

cladmi committed:
pkg/fatfs: Cleanup to use RIOTPKG variable

cladmi committed:
pkg: protect all Makefile.dep dependencies

Allow to safely include all pkg/*/Makefile.dep before knowing if they are used
or not.

This allows using "USEMODULE += pkg_name_submodule_name" in applications and
define the 'pkg_name_submodule_name' dependency to 'pkg_name' in the package
itself.

cladmi committed:
Makefile.dep: include all packages Makefile.dep

This allows packages to declare there sub modules dependencies directly in the
package directory instead of the main Makefile.dep.

It however requires all packages to protect Makefile.dep to only set
dependencies if the package is used.

cladmi commented:
The incremental changes also allow to possibly stop on a previous commit if preferred, or split this PR.

cladmi commented:
I also remember there are some packages pre-cleanup included that I can do in sub PRs.

cladmi committed:
makefiles/info-global.inc.mk: remove mentions of USEPKG

cladmi commented on Makefile.include:
```diff
@@ -250,9 +250,22 @@ ifneq ($(GNRC_NETIF_NUMOF),1)
 
+# Detect if an application or Makefile.dep defined USEPKG
+ifneq (,$(USEPKG))
+  ifeq ($(origin USEPKG), file)
```
I currently need this because USEPKG is exported in `vars.inc.mk` and it seems already set before executing make in murdock.

I could remove the export in `vars.inc.mk` maybe, but I still do not get why it has a value when passing through here for the moment (I did not investigate).

cladmi replied:
There is currently a usage in `sys/shell/commands` that should be updated to use `USEMODULE`. I will try to go early in that direction.

cladmi commented:
I fixed the broken builds, remaining task is to move packages dependencies out of the main Makefile.dep.

cladmi commented:
Still remaining dependencies not migrated to `pkg/*/Makefile.dep`
They may be done later too.

* lwip%
  ```
  ifneq (,$(filter ieee802154 nrfmin,$(USEMODULE)))
    ...
    ifneq (,$(filter lwip%, $(USEMODULE)))
      USEMODULE += lwip_sixlowpan
    endif
  endif
  ```
* nordic_softdevice_ble removing modules
  ```
  ifneq (,$(filter nordic_softdevice_ble,$(USEMODULE)))
    # prevent application from being a router
    # TODO: maybe find a nicer solution in future build system update
    _ROUTER_MODULES = gnrc_ipv6_router% gnrc_rpl netstats_rpl auto_init_gnrc_rpl
    ifneq (,$(filter $(_ROUTER_MODULES),$(USEMODULE)))
      $(warning nordic_softdevice_ble: Disabling router modules:\
        $(filter $(_ROUTER_MODULES),$(USEMODULE)))
    endif
    USEMODULE := $(filter-out $(_ROUTER_MODULES),$(USEMODULE))
  endif
  ```
* openthread dependencies which are now all defined in the test Makefile

gebart commented:
There's a weird lwip failure in the Murdock logs. Maybe needs a rebase?

cladmi commented:
I moved the `lwip` dependencies to `pkg/lwip/Makefile.dep` and re-order them so it could also be my fault.

cladmi commented:
Looks still broken, I will investigate. I expected this as some of the dependencies definition are order specific.
If I do not manage easily, I will keep the problematic `lwip` ones in the main `Makefile.dep`

cladmi commented:
I moved the "default implementation" selection at the end of `pkg/lwip/Makefile.dep` and it should be fixed with this.

cladmi commented:
Will need to rebase due to #9884

smlng commented:
@cladmi how is this coming? I think this is a nice cleanup, if you need help on testing: I'm in.

cladmi commented:
Thanks for the update. It got somehow lost in my todo list, and was waiting for the "Get maintainers feedback on the solution" and "Check if the updated documentation is enough ! ! !" parts so did not put more work on it recently.

If you like this current solution, or tell me what I should change, I can start splitting parts.

cladmi committed:
pkgs: move packages dependencies to pkg/*/Makefile.dep

Move most of the packages dependencies to pkg/*/Makefile.dep.
Still some weird cases are handled in Makefile.dep.

I used this command to check for non migrated packages

    grep $(ls -d pkg/*/ | sed 's|^pkg/|-e |;s|/||') Makefile.dep

cladmi commented:
I did a blind stupid rebase on master. May not be working right now.

From the first version I did, I would like to change the handling to do the dependencies first to allow deprecating incompatible things earlier without breaking anything.

I see some steps to go forward with this in parallel:

* [ ] cleanups
* [ ] Declare `ALL_PKGS`
* [ ] Always add `USEPKG` to `USEMODULE`
  * [ ] change the usages of `$(USEPKG)` to `$(USEMODULE)`.
    * [ ] do not export `USEPKG` anymore
  * [ ] Change the packages `Makefile.dep` to be always with `ifneq(,$(filter packagename,$(USEMODULE))` and add a sanity check for it (this will allow deprecating it and not break external code immediately).
    * [ ] With this, the packages `Makefile.dep` could always be parsed without side-effect, like done for all the others `Makefile.dep` right now.
